[System.Drawing] Add unit tests based on Mono's test suite

qmfrederik opened the pull request:
When consolidating the Unix and Windows codebase for System.Drawing, I noticed some gaps in code coverage. For example, the various `Graphics.*Clip*` methods and properties did not have any coverage.

As a result, I tried to port the Mono test suite for System.Drawing to corefx. This PR is the result of that.

The tests uncovered a couple of bugs in the current implementation of System.Drawing:
-  The getter for `PathGradientBrush.Blend` returned inconsistent results across calls due to invalid P/Invoke code -  https://github.com/dotnet/corefx/commit/3a6f13043b6790449cc2c0b1d4c44b3c7e15bc9b
- The P/Invoke declaration for `GdipSaveImageToFile_delegate` was missing a `MarshalAs` attribute for the `filename` parameter, causing the filenames to be passed in ANSI format to a function which expected Unicode, resulting in unexpected filenames.
- The `InvalidEnumArgumentException` exception class exists in both `System.Drawing` and `System.ComponentModel`

qmfrederik commented:
/cc @mellinoe @hughbe 

In general, I think it's worth adding these tests to the test suite for System.Drawing as they were written when Mono's System.Drawing was written, and were used to assert that Mono's System.Drawing exposed the same behavior as the on on .NET.

In addition to that, they have the advantage of testing the System.Drawing + GDI+ layer, and tend to be "high level" tests of the framework. For example, lots of these tests load images or draw lines, arcs,... and then do a near-pixel-level validation of the result.

I'm aware there's still a lot of work to be done on these tests to get them in a fairly stable state, that would include
- Code formatting standards
- Storing blobs used as test asset in a NuGet package rather than git
- ...

It would make sense to conslidate these unit tests. A way forward would be to finish & merge this PR first, and then do the consolidation in separate PRs, just like we are doing for the actual codebase.

Thoughts?

ViktorHofer commented:
> Storing blobs used as test asset in a NuGet package rather than git

Once you commit sth into a git repository the files will stay there forever (because of history). If the files which are used here are large then I definitely suggest putting them into the accompanying repo for blobs before merging.

qmfrederik commented:
The Linux builds fail with this error:

```
09:48:15 CSC : error CS1504: Source file '/mnt/j/workspace/dotnet_corefx/master/linux-TGroup_netcoreapp+CGroup_Release+AGroup_x64+TestOuter_false_prtest/src/System.Drawing.Common/tests/mono/System.Drawing.Imaging/IconCodecTest.cs' could not be opened
 -- Unable to load DLL 'api-ms-win-core-localization-l1-2-0.dll':
 The specified module or one of its dependencies could not be found. 
[/mnt/j/workspace/dotnet_corefx/master/linux-TGroup_netcoreapp+CGroup_Release+AGroup_x64+TestOuter_false_prtest/src/System.Drawing.Common/tests/System.Drawing.Common.Tests.csproj]
```

Does anyone know why CSC would try to load `api-ms-win-core-localization-l1-2-0.dll` on Linux?

qmfrederik commented:
@dotnet-bot test Windows x64 Debug Build

qmfrederik commented:
OK - the tests are passing on Windows which seems like a good start.

For Linux, some tests fail on some distros only: perhaps they have different versions of libgdiplus?
For Linux tests that fail on all Linux distros: these may be actual regressions on Linux as a result of the consolidation push. I'll have a closer look.

akoeplinger commented:
> For Linux, some tests fail on some distros only: perhaps they have different versions of libgdiplus?

If I remember correctly, at least the `Bitmap2bit*` and `Save_*` tests were due to some bugfix in newer libgdiplus.

One idea I had about how to handle cases like these would be to add an API to libgdiplus which returns the version numbers, so we could skip tests on broken versions.

qmfrederik commented:
> One idea I had about how to handle cases like these would be to add an API to libgdiplus which returns the version numbers, so we could skip tests on broken versions.

+1 for this one, we could also consider using it to prevent some code from executing if we know it would trigger issues in the underlying libgdiplus implementation.

akoeplinger commented on src/System.Drawing.Common/tests/mono/System.Drawing/tests.sh:
```diff
@@ -0,0 +1,29 @@
+#!/bin/sh
```
this file can be deleted

akoeplinger commented on src/System.Drawing.Common/tests/mono/System.Drawing/tests-ms.sh:
```diff
@@ -0,0 +1,28 @@
+#!/bin/sh
```
this file can be deleted

akoeplinger commented on src/System.Drawing.Common/tests/mono/System.Drawing.Imaging/.gitattributes:
```diff
@@ -0,0 +1,2 @@
+/TestColorMatrix.cs -crlf
```
can probably be deleted as well

akoeplinger commented on src/System.Drawing.Common/tests/mono/System.Drawing/.gitattributes:
```diff
@@ -0,0 +1 @@
+/tests-ms.sh -crlf
```
delete

akoeplinger commented on src/System.Drawing.Common/tests/mono/System.Drawing/README:
```diff
@@ -0,0 +1,12 @@
+created by tests in the directory called MsNet.
+
+
+
```
can be deleted

qmfrederik commented:
@akoeplinger Thanks! - I removed those files.

qmfrederik commented:
[According to pkgs.org](https://pkgs.org/download/libgdiplus), CentOS 7.3 and RHEL 7.3 are on version 2.10 of libgdiplus; Ubuntu 14.04 uses version 2.11.

[libgdiplus 2.10](https://github.com/mono/libgdiplus/commits/2.10) dates from Jan 13, 2011 (with the last commit on Oct 15, 2010), so that seems rather outdated :-).

@mellinoe Any way we can double-check which version of libgdiplus is installed on the build servers? The Mono project hosts package feeds for libgdiplus for most Linux distro's, perhaps it makes sens to use libgdiplus from that feed?

qmfrederik commented:
Current status: 84 unit test failures. A fair deal of them are on systems which I suspect are running old versions of libgdiplus (CentOS, RHEL and Ubuntu 14.04) - I'll go ahead & try to disable these tests on those systems.

qmfrederik commented:
The last build saw 3 unit test failures in ` System.Drawing.Tests.BrushesTests/Brushes_Get_ReturnsExpected` for `YellowGreen`. It appears Mono's unit tests actually break `YellowGreen` by a destructive test; I'll fix that.

The fixes for bugs uncovered by these unit tests have landed separately - via #23956, #23957 and #23958.

Work to keep the blobs out of git is also on it's way.

qmfrederik commented:
@dotnet-bot test code coverage please

mellinoe commented on src/System.Drawing.Common/src/System.Drawing.Common.csproj:
```diff
@@ -34,13 +34,13 @@
     <Reference Include="System.Runtime.InteropServices" />
     <Reference Include="System.ComponentModel" />
     <Reference Include="System.ComponentModel.Primitives" />
+    <Reference Include="System.ComponentModel.TypeConverter" />
```
We actually don't want this dependency. It was previously there and we removed it due to layering concerns.

mellinoe replied:
And you are correct that we have duplicate `InvalidEnumArgumentException` types, but that is going to be okay for now. The one defined in System.Drawing.Common is internal, and the real type is being moved down into System.ComponentModel.Primitives, but that won't be available until .NET Core 2.1. In the meantime, we are building against .NET Core 2.0 and will just include an internal version of the exception.

qmfrederik replied:
OK, I'll revert the change. It causes one unit test to fail, I'll disable that test for now.

qmfrederik commented:
@mellinoe & others, quick update:

- I removed the change for `InvalidEnumArgumentException` and disabled the tests which check for this exception.
- There's one bitmap file missing in the test data NuGet package, I sent out a PR for that.
- Other than that, I have hopes that we can stabilize this PR this week, as in all unit tests passing.

What else do you need for this PR to merged? From a code style standpoint, I haven't change much yet to keep the diff with the upstream Mono tests fairly simple, but once the tests are stable I can tend to that.
Other than opening each file in VS and having VS auto-format the document (*), anything else to watch out for?

(*): When I make that change, I'll remember all the spaces I carefully put between the function name and the argument list, as well as the conversions from spaces to tabs and the countless times @akoeplinger had to remind me of Mono's code style ;-)

qmfrederik commented:
@mellinoe Thanks, I've bumped the System.Drawing test data package, I'll keep an eye on the CI results as they come in.

mellinoe commented:
> Other than opening each file in VS and having VS auto-format the document (*), anything else to watch out for?

That will be a good start. Realistically, I don't care as much about the other stuff like variable names, etc. but the braces and tabs will be problematic if anyone formats the document with VS. Can we do that before we merge?

qmfrederik commented:
@mellinoe Yup, I'll make sure all documents use the default Visual Studio code formatting. I'll do it once the CI is green, shouldn't be long before we get there.

qmfrederik commented:
@mellinoe Two quick questions:
- The NETFX x86 Release Build tests fail, but I don't seem to be able to see detailed test results (as in, which tests failed). Is there any way to retrieve them?
- It also seems the System.Drawing tests are not running on macOS. Do you know whether libgdiplus has been installed on the macOS build servers?

It looks like the System.Threading.Tasks.Tests crashed on suse.422.amd64.Open, I'll relaunch the Linux tests.

qmfrederik commented:
@dotnet-bot test Linux x64 Release Build please

mellinoe commented:
> It also seems the System.Drawing tests are not running on macOS. Do you know whether libgdiplus has been installed on the macOS build servers?

Yeah, I just noticed this in another PR as well. It is supposed to be installed on them via homebrew. I'll see if I can figure out what's wrong.

mellinoe commented:
You should be able to see the full output of the netfx tests here:

https://ci3.dot.net/job/dotnet_corefx/job/master/job/windows-TGroup_netfx+CGroup_Release+AGroup_x86+TestOuter_false_prtest/3411/consoleText

There are quite a few tests that are failing in Drawing. If any of the mono tests are asserting behavior that doesn't match the .NET Framework behavior, we need to remove those tests.

qmfrederik commented:
@mellinoe Thanks, some of the results are indeed consistent between Desktop & Core (oddly enough, also on Windows) so I disabled those tests for now.
Other tests were checking for ArgumentExceptions where the Core implementation has been changed to throw an ArgumentNullException. I've reverted that change for the time being (9c37c93) to keep the behavior consistent.

safern commented:
`MonoTests.System.Drawing.IconTest.Constructor_StreamNull`
`MonoTests.System.Drawing.IconTest.Constructor_StreamNull_Size`

Are checking for `ArgumentNullException` also.

The remaining failures in NETFX are in `PathGradientBrushTest` class. You can repro them locally by running: 
```
build.cmd -framework:netfx
msbuild <csproj> /t:rebuildandtest /p:targetgroup=netfx
```

qmfrederik commented:
@dotnet-bot test code coverage please

qmfrederik commented:
@mellinoe @safern All unit tests passed (finally!), and I ran code formatting on the newly added files. Let me know if there's anything else you need.

mellinoe commented:
@qmfrederik Are you able to test on a Mac? There are a couple of issues blocking the OSX CI at the moment, but it would be good if you could run these locally (with my other PR #24123 ) and make sure everything works. If you don't have an OSX machine then I can borrow one.

mellinoe commented:
Sorry, I missed this part:

> Other tests were checking for ArgumentExceptions where the Core implementation has been changed to throw an ArgumentNullException. I've reverted that change for the time being (9c37c93) to keep the behavior consistent.

Is that really necessary? We did that because we consider throwing a derived exception to be a non-breaking change, and `ArgumentNullException` more closely matches what is happening in those cases. Can we just fix up the mono tests slightly to allow for it?

safern commented:
> Can we just fix up the mono tests slightly to allow for it?

You can fix it with https://github.com/xunit/assert.xunit/blob/master/ExceptionAsserts.cs#L70 asserting for `ArgumentException`

I agree on keeping the `ArgumentNullException` as it derives from `ArgumentException`

qmfrederik commented:
@mellinoe @safern I rebased off master, so that should bring the OS X fix in this branch, and reverted the changes for the `ArgumentException`.

I do have access to a Mac, so I can test on one if needed.

qmfrederik commented:
@dotnet-bot test code coverage please

qmfrederik commented:
@mellinoe @safern We're getting close :). All tests are green (even on macOS); the `Argument(Null)Exception` situation has been cleared up, and code formatting is in place.

Let me know if you need anything else.

qmfrederik commented:
Rebased now that #24140 got applied, this PR now only contains changes to the test code and no product changes.

mmitche commented:
@dotnet-bot test this please (testing)

qmfrederik commented:
Looks like the NETFX x86 release crashed & is hanging as a result:

```
09:57:18 ERROR: Execution failed
09:57:18 java.lang.InterruptedException
09:57:18 	at java.lang.Object.wait(Native Method)
09:57:18 	at hudson.remoting.Request.call(Request.java:147)
09:57:18 	at hudson.remoting.Channel.call(Channel.java:829)
09:57:18 	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler.execute(RemoteGitImpl.java:146)
09:57:18 	at sun.reflect.GeneratedMethodAccessor504.invoke(Unknown Source)
09:57:18 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
09:57:18 	at java.lang.reflect.Method.invoke(Method.java:498)
09:57:18 	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler.invoke(RemoteGitImpl.java:132)
09:57:18 	at com.sun.proxy.$Proxy179.execute(Unknown Source)
09:57:18 	at hudson.plugins.git.GitSCM.retrieveChanges(GitSCM.java:1075)
09:57:18 	at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1115)
09:57:18 	at org.jenkinsci.plugins.workflow.steps.scm.SCMStep.checkout(SCMStep.java:109)
09:57:18 	at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:83)
09:57:18 	at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:73)
09:57:18 	at org.jenkinsci.plugins.workflow.steps.AbstractSynchronousNonBlockingStepExecution$1$1.call(AbstractSynchronousNonBlockingStepExecution.java:47)
09:57:18 	at hudson.security.ACL.impersonate(ACL.java:260)
09:57:18 	at org.jenkinsci.plugins.workflow.steps.AbstractSynchronousNonBlockingStepExecution$1.run(AbstractSynchronousNonBlockingStepExecution.java:44)
09:57:18 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
09:57:18 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
09:57:18 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
09:57:18 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
09:57:18 	at java.lang.Thread.run(Thread.java:748)
```

danmosemsft commented:
@dotnet/dnceng is the InterruptedException above caused by the official build issue going on?

markwilkie commented:
Only VSTS official builds are affected by the hotfix that broke our builds.

qmfrederik commented:
@dotnet-bot test NETFX x86 Release Build please

mmitche commented:
@danmosmsft This might have been caught by my restart this morning.  Let me know if you see it again.

mellinoe commented on src/System.Drawing.Common/tests/mono/System.Drawing.Drawing2D/GraphicsPathIteratorTest.cs:
```diff
@@ -0,0 +1,170 @@
+//
```
We need to include a 2-line header in all of these new files.

```C#
// Licensed to the .NET Foundation under one or more agreements.
// See the LICENSE file in the project root for more information.
```

qmfrederik replied:
Just to be clear, do you need this header in addition to the existing header or does it replace the existing header?

mellinoe replied:
It is in addition to the existing header -- it just goes at the very top.

mellinoe replied:
The files under System.Drawing.Common/src/Unix show what it should look like. Those are files pulled in from the mono codebase.

mellinoe commented on src/System.Drawing.Common/tests/mono/System.Drawing.Drawing2D/GraphicsPathIteratorTest.cs:
```diff
@@ -0,0 +1,170 @@
+        [ConditionalFact(Helpers.GdiplusIsAvailable)]
+        public void Ctor_Null()
+        {
+            using (GraphicsPathIterator gpi = new GraphicsPathIterator(null))
```
This is identical to an existing test that we already have checked in

https://github.com/dotnet/corefx/blob/master/src/System.Drawing.Common/tests/Drawing2D/GraphicsPathIteratorTests.cs#L61

mellinoe commented on src/System.Drawing.Common/tests/mono/System.Drawing.Drawing2D/GraphicsPathIteratorTest.cs:
```diff
@@ -0,0 +1,170 @@
+        }
+
+        [ConditionalFact(Helpers.GdiplusIsAvailable)]
+        public void NextMarker_Null()
```
This is identical to an existing test.

mellinoe commented on src/System.Drawing.Common/tests/mono/System.Drawing.Drawing2D/GraphicsPathIteratorTest.cs:
```diff
@@ -0,0 +1,170 @@
+        }
+
+        [ActiveIssue(20844)]
+        public void NextSubpath_Null()
```
This is identical to an existing test

mellinoe commented on src/System.Drawing.Common/tests/mono/System.Drawing.Drawing2D/GraphicsPathIteratorTest.cs:
```diff
@@ -0,0 +1,170 @@
+        }
+
+        [ConditionalFact(Helpers.GdiplusIsAvailable)]
+        public void CopyData_NullPoints()
```
This is identical to an existing test.

mellinoe commented on src/System.Drawing.Common/tests/mono/System.Drawing.Drawing2D/GraphicsPathIteratorTest.cs:
```diff
@@ -0,0 +1,170 @@
+        }
+
+        [ConditionalFact(Helpers.GdiplusIsAvailable)]
+        public void CopyData_NullTypes()
```
This is basically covered by an existing test. https://github.com/dotnet/corefx/blob/master/src/System.Drawing.Common/tests/Drawing2D/GraphicsPathIteratorTests.cs#L351

mellinoe commented:
@qmfrederik I have started to look through the test code, and I'm concerned that there is a very large amount of overlap with existing tests. Before adding 30,000+ lines of code, we should be sure that it's adding real value to un-tested parts of the library. Have we done any de-duplication with what exists in corefx already? A big part of the problem here is that a majority of the test cases we have are already ported from mono's codebase -- just significantly cleaned up and improved. I've looked at the `GraphicsPathIterator` tests and I believe that most or all of the test cases are duplicates. In the cases I looked at, the code we have in corefx is better formatted and has more exhaustive coverage, as well.

I know this is a difficult problem, so I'm not sure how best to proceed. I think 30,000 lines of test code is probably too much for a single pull request, especially when there is such duplication. Can we split this up into a few individual PR's and try to do some de-duplication along the way?

danmosemsft commented:
Aside from @mellinoe 's feedback, note that our license header would need merging in to each of these files before the change could be merged. I'll mark it no merge for now, for that reason.

qmfrederik commented:
@mellinoe Sure, we can split this up in multiple PRs. As to the value - yes, I believe there is value in this; #24140, #23958, #23957, #23956 for example are fixes for regressions that these tests caught.

My only concern is that some of the refactoring/consolidation we're doing impacts the codebase without us noticing (see the list of PRs before). On the other hand, we can keep this PR open for the time being - it would catch regressions, if any, whenever this is rebased.

In general, I think the Mono's test have the advantage of also testing the GDI+ code and how this library interacts with it.

mellinoe commented:
@qmfrederik I definitely agree -- there's value in here, and you've already found and fixed issues because of it. I'd just like to do at least a basic scan of each class to make sure there's not excessively duplicated stuff. I think a little bit of overlap will be fine; like you mentioned we did a lot of refactoring so some of the new tests are not strictly equivalent to the old tests. Looking at the `GraphicsPathIterator` tests, I was able to spot a lot of duplicates just by looking at the method names. I think just a simple scan like that should be good enough to get rid of most of the duplication, and will leave us with a more manageable test bed.

If we separate this into a few PR's, I'd be happy to take one of the namespaces and do the de-duplication from your branch. That way we could split up the extra work I'm creating by asking for this 😄

qmfrederik commented:
Sounds good :-). There are some fairly large files in there - Graphics and Bitmap come to mind. I'll split those off (as well as Metafile, which has very little code coverage at this point) and we can start with those.

Feel free to leave comments on the PR, send a PR to my branch, fork & create a separate PR, ... .

mellinoe commented:
@qmfrederik I'm going to work on pulling out the Printing tests from here.

mellinoe commented:
Actually, I looked through all of the Printing tests, and I don't think there's anything valuable there. Everything is a duplicate of what we already have, except for `PrintingServicesUnixTest.cs`. That test might be useful, but it won't work as-is (it uses direct PInvokes to libcups).

mellinoe commented:
@qmfrederik As noted in #24263 , I looked at both of the System.Drawing.Text files you have here (there are only two) and found that our coverage is equal or better. I put up that PR to re-enable our existing tests everywhere.

qmfrederik commented:
Thanks for letting me know! I'll remove that file from this PR - may help as a kind of bookkeeping as to which tests are redundant and which test we may want to consider porting over.

mellinoe commented:
@qmfrederik Yep, sounds good. I'm trying to pick off the outer stuff since it will be quick and I don't want to conflict with your current PR's.

I also looked at `ImageFormatTests.cs`, and it is also duplicate coverage. Our version is pretty much identical in function, just refactored a little bit to use `[Theory]`.

mellinoe commented:
`WmfPlaceableFileHeaderTest.cs` is the same thing as our `tests/Imaging/WmfPlaceableFileHeaderTests.cs`, ours is just refactored slightly (again with `[Theory]`).

karelz commented:
@safern what is the remaining work here?

safern commented:
> @safern what is the remaining work here?

@qmfrederik splitted this PR into multiple PRs to find test duplication and make reviewing easier. So I will just wait for his confirmation that we can close this PR.

karelz commented:
In that case I would recommend to close it right away as we do not plan to merge it ever. The content can be easily harvested and reused even if it is closed ... @qmfrederik what do you think?

safern commented:
This is why we kept it open: https://github.com/dotnet/corefx/pull/23797#issuecomment-330653568

karelz commented:
> we can keep this PR open for the time being - it would catch regressions, if any, whenever this is rebased

Is anyone planning to do that anytime soon? If yes, sure ... if not, let's close it.

qmfrederik commented:
I just rebased to see whether these tests are still in a good shape ;-).

I'll try to go over the PR again this weekend/early next week to see if there are obvious tests that we can split off in separate PRs or if it's really just duplicates at this point and we can close the PR.

danmosemsft commented:
Thanks @qmfrederik we appreciate your work here.

qmfrederik commented:
Closing for now, I hope to be able to come back to this in the near future.
